lib/awful/widget/watch.lua

```diff
@@ -0,0 +1,79 @@
+---------------------------------------------------------------------------
+--- Watch widget.
+--
+-- @release @AWESOME_VERSION@
+-- @classmod awful.widget.watch
+---------------------------------------------------------------------------
+
+local setmetatable = setmetatable
+local textbox = require("wibox.widget.textbox")
+local timer = require("gears.timer")
+local spawn = require("awful.spawn")
+
+local watch = { mt = {} }
+
+local function easy_async_spawn(cmd, callback)
+  local stdout = ''
+  local stderr = ''
+  local function parse_stdout(str)
+    stdout = stdout .. str .. "\n"
+  end
+  local function parse_stderr(str)
+    stderr = stderr ..  str .. "\n"
+  end
+  local function done_callback()
+    return callback(stdout, stderr)
+  end
+  local exit_callback_fired = false
+  local output_done_callback_fired = false
+  local function exit_callback()
+    exit_callback_fired = true
+    if output_done_callback_fired then
+      return done_callback()
+    end
+  end
+  local function output_done_callback()
+    output_done_callback_fired = true
+    if exit_callback_fired then
+      return done_callback()
+    end
+  end
+  return spawn.with_line_callback(
+    cmd, {
+    stdout=parse_stdout,
+    stderr=parse_stderr,
+    exit=exit_callback,
+    output_done=output_done_callback
+  })
```

Comment:
Since my comments elsewhere got lost:

You do not need an `exit_callback`, just `output_done` is enough for this code (that simplifies the code a bit).